### ISSUE TITLE
[Feature-Request] WBF(Weighted Boxes Fusion) over NMS(Non-maximum Suppression)

### ISSUE THREAD

**Issue by dsbyprateekg** [david8862/keras-YOLOv3-model-set#100]
Hi @david8862 ,

Recently working on [https://www.kaggle.com/c/global-wheat-detection/overview](url) challenge, I see high scorers are using WBF over the NMS with ultralytics's YOLO5 and EfficientDet which gives them a good boost in their scores. So I thought to present this with you.

The result of different approaches with WBF on Open Images Dataset is as below-
![wbf_map](https://user-images.githubusercontent.com/30830541/88135455-b6618000-cc04-11ea-94eb-9e9beb8eb6dc.JPG)

Link of the paper is [https://arxiv.org/abs/1910.13302](url)

Python implementation is [https://github.com/ZFTurbo/Weighted-Boxes-Fusion/blob/master/ensemble_boxes/ensemble_boxes_wbf.py](url)

If you find it suitable, please add this approach. I believe it will boost YOLOv4 score in competitions.

**Comment by david8862** [comment]
@dsbyprateekg  many thanks for the info. I'll check and try to integrate into code base, but may not in time due to tasks on other project.

**Comment by dsbyprateekg** [comment]
Thanks @david8862 
There is no hurry, please take your time to add this.

**Comment by david8862** [comment]
@dsbyprateekg  I've ported WBF into numpy postprocess module, but seems it didn't have better performance than NMS on single YOLO model bbox handling. There're also related discussion at https://github.com/ZFTurbo/Weighted-Boxes-Fusion/issues/2

**Comment by dsbyprateekg** [comment]
@david8862 Thanks for looking into this.
It seems for single object detection and recognition model it is not performing well.
I can see that in current challenge of wheat head detection in Kaggle.
People are using **WBF with TTA** in their single model. Here are some examples-

[https://www.kaggle.com/wasupandceacar/yolov5-single-model-with-more-tta-lb-0-745](url)

[https://www.kaggle.com/c/global-wheat-detection/discussion/156489](url)

I will try to add WBF with TTA in my model and then I will share the result.

**Comment by dsbyprateekg** [comment]
Hi @david8862 ,

I have done the experiment with WBF for the current ongoing Kaggle challenge and found that my score is improved from 0.6626 (with NMS) to 0.6815 (with WBF).

Although when I compared an image with NMS and with WBF, I found that confidence score with NMS is far better than with WBF. See below image with bbox and score-

Image with NMS:
![51b3e36ab](https://user-images.githubusercontent.com/30830541/88451210-97682580-ce72-11ea-9dff-734a0838992b.jpg)

Image with WBF:
![51b3e36ab](https://user-images.githubusercontent.com/30830541/88451221-a8b13200-ce72-11ea-81f5-2f784978670c.jpg)

This contradiction may be due to the evaluation criteria of this specific Kaggle challenge. For this challenge, I will stick to WBF but for other problems, I will stick to NMS.

Thanks again for your reply and for adding this to your repository!
